working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+`dynamic`.
+
+It is an error for an instance field with no static initializer to have a
+potentially nullable type unless the variable or field is marked with the `lazy`
+modifier.
+
+It is an error if a local variable that is potentially non-nullable and is not
+marked `lazy` is used before it is definitely assigned (see Definite Assignment
+below).
+
+It is an error if a method, function, getter, or closure with a potentially
+non-nullable return type does not definitely complete (see Definite Completion
+below).
+
+It is an error if an optional parameter (named or otherwise) with no default
+value has a potentially non-nullable type.
+
+It is an error if a required named parameter has a default value.
+
+It is an error if a named parameter that is part of a `required` group is not
+bound to an argument at a call site.
+
+It is an error to call the default `List` constructor with a length argument and
+a type argument which is potentially non-nullable.
+
+For the purposes of errors and warnings, the null aware operators `?.` and `?..`
+are checked as if the receiver of the operator had non-nullable type.
+
+It is an error for a class to extend, implement, or mixin a type of the form
+`T?` for any `T`.
+
+It is an error for a class to extend, implement, or mixin the type `Never`.
+
+It is an error to call a method, setter, or getter on a receiver of static type
+`Never` (including via a null aware operator).
+
+It is an error to apply an expression of `Never` in the function position in a
+function call.
+
+It is an error if the static type of `e` in the expression `throw e` is
+potentially nullable.
+
+It is not an error for the body of a `lazy` field to reference `this`.
+
+It is an error for a toplevel variable, a static field, or a formal parameter to
+be declared `lazy`.
+
+It is an error if the type `T` in the **on-catch** clause `on T catch` is
+potentially nullable.
+
+### Assignability
+
+The definition of assignability is changed as follows.  
+
+A type `T` is **assignable** to a type `S` if `T` is `dynamic`, or if `S` is a
+subtype of `T`.
+
+### Generics
+
+The default bound of generic type parameters is treated as `Object?`.
+
+### Type promotion, Definite Assignment, and Definite Completion
+
+**TODO** Fill this out.
+
+### Null promotion
+
+The machinery of type promotion is extended to promote the type of variables
+based on nullability checks subject to the same set of restrictions as normal
+promotion.  The relevant checks and the types they are considered to promote to
+are as follows.
+
+A check of the form `e == null` or of the form `e is Null` where `e` has static
+type `T` promotes the type of `e` to `Null` in the `true` continuation, and to
+**NonNull**(`T`) in the
+`false` continuation.
+
+A check of the form `e != null` or of the form `e is T` where `e` has static
+type `T?` promotes the type of `e` to `T` in the `true` continuation, and to
+`Null` in the `false` continuation.
```

Comment:
We might also want:

A check of the form `e is T` where `e` has static type `FutureOr<T>` promotes `e` to `T` in the `true` continuation, and to `Future<T>` in the `false` continuation.

(Or is this just assumed part of the TODO above, since it's not null-related?)

Reply:
Yes, part of TODO.